homeassistant/components/somfy/__init__.py

```diff
@@ -0,0 +1,160 @@
+
+
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry):
+    """Set up Somfy from a config entry."""
+    def token_saver(token):
+        _LOGGER.debug('Saving updated token')
+        entry.data[CONF_TOKEN] = token
+        hass.config_entries.async_update_entry(entry, data={**entry.data})
+
+    # Force token update.
+    from pymfy.api.somfy_api import SomfyApi
+    hass.data[DOMAIN][API] = SomfyApi(
+        entry.data['refresh_args']['client_id'],
+        entry.data['refresh_args']['client_secret'],
+        token=entry.data[CONF_TOKEN],
+        token_updater=token_saver
+    )
+
+    await update_all_devices(hass)
+
+    for component in SOMFY_COMPONENTS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component))
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistantType, entry: ConfigEntry):
+    """Unload a config entry."""
+    if not hass.data[DOMAIN]:
+        hass.data.pop(DOMAIN)
+
+    for component in SOMFY_COMPONENTS:
+        await hass.config_entries.async_forward_entry_unload(entry, component)
+
+    return True
+
+
+class SomfyEntity(Entity):
+    """Representation of a generic Somfy device."""
+
+    def __init__(self, device, api):
+        """Initialize the Somfy device."""
+        self.device = device
+        self.api = api
+
+    @property
+    def unique_id(self):
+        """Return the unique id base on the id returned by Somfy."""
+        return self.device.id
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self.device.name
+
+    @property
+    def device_info(self):
+        """Return device specific attributes.
+
+        Implemented by platform classes.
+        """
+        return {
+            'identifiers': {(DOMAIN, self.unique_id)},
+            'name': self.name,
+            'model': self.device.type,
+            'via_hub': (DOMAIN, self.device.site_id),
+            # For the moment, Somfy only returns their own device.
+            'manufacturer': 'Somfy'
+        }
+
+    async def async_update(self):
+        """Update the device with the latest data."""
+        await update_all_devices(self.hass)
+        devices = self.hass.data[DOMAIN][DEVICES]
+        self.device = next((d for d in devices if d.id == self.device.id),
+                           self.device)
+
+    def has_capability(self, capability):
+        """Test if device has a capability."""
+        capabilities = self.device.capabilities
+        return bool([c for c in capabilities if c.name == capability])
+
+
+@Throttle(MIN_TIME_BETWEEN_UPDATES)
+async def update_all_devices(hass):
+    """Update all the devices."""
+    from requests import HTTPError
+    try:
+        data = hass.data[DOMAIN]
+        data[DEVICES] = await hass.async_add_executor_job(
+            data[API].get_devices)
+    except HTTPError:
+        _LOGGER.warning("Cannot update devices", exc_info=True)
```

Comment:
Why print a full strack trace of the exception ? That seems a bit much. Why not just include the stringied version of the exception ?

Reply:
I like to log full stack trace, it help to debug and understand what's wrong more than the name of the exception. Moreover, it should not happen so many times.

Reply:
Don't log the full stack trace. This is a simple connection error with a unique log message. The stack trace adds 0 value. Remove it.